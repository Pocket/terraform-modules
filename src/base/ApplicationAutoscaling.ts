import { Resource } from 'cdktf';
import { appautoscaling, cloudwatch, iam } from '@cdktf/provider-aws';
const { AppautoscalingPolicy, AppautoscalingTarget } = appautoscaling;
const { CloudwatchMetricAlarm } = cloudwatch;
const { IamRole, IamRolePolicy, DataAwsIamPolicyDocument } = iam;
import { Construct } from 'constructs';

export interface ApplicationAutoscalingProps {
  ecsClusterName: string;
  ecsServiceName: string;
  prefix: string;
  scalableDimension: string;
  scaleInThreshold: number;
  scaleOutThreshold: number;
  stepScaleInAdjustment: number;
  stepScaleOutAdjustment: number;
  tags?: { [key: string]: string };
  targetMaxCapacity: number;
  targetMinCapacity: number;
}

/*
 * Generates an AutoScaling group
 */

export class ApplicationAutoscaling extends Resource {
  constructor(
    scope: Construct,
    name: string,
    config: ApplicationAutoscalingProps
  ) {
    super(scope, name);

    // set up IAM role & policy
    const iamRole = ApplicationAutoscaling.generateIamRole(this, config);
    ApplicationAutoscaling.generateIamRolePolicy(this, config, iamRole);

    // set up autoscaling target & in/out policies
    const autoScalingTarget = ApplicationAutoscaling.generateAutoScalingTarget(
      this,
      config,
      iamRole
    );

    const applicationScaleOut =
      ApplicationAutoscaling.generateAutoSclaingPolicy(
        this,
        config,
        autoScalingTarget,
        'Out'
      );

    const applicationScaleIn = ApplicationAutoscaling.generateAutoSclaingPolicy(
      this,
      config,
      autoScalingTarget,
      'In'
    );

    // set up cloudwatch alarms
    ApplicationAutoscaling.generateCloudwatchMetricAlarm(
      this,
      config,
      'scale_out_alarm',
      `${config.prefix} Service High CPU`,
      'Alarm to add capacity if container CPU is high',
      'GreaterThanThreshold',
      config.scaleOutThreshold,
      applicationScaleOut.arn
    );

    ApplicationAutoscaling.generateCloudwatchMetricAlarm(
      this,
      config,
      'scale_in_alarm',
      `${config.prefix} Service Low CPU`,
      'Alarm to reduce capacity if container CPU is low',
      'LessThanThreshold',
      config.scaleInThreshold,
      applicationScaleIn.arn
    );
  }

  /**
   * Creates an IAM Role
   * @param resource
   * @param config
   * @returns IamRole
   */
  static generateIamRole(
    resource: Resource,
    config: ApplicationAutoscalingProps
  ): iam.IamRole {
    return new IamRole(resource, `autoscaling_role`, {
      name: `${config.prefix}-AutoScalingRole`,
      assumeRolePolicy: new DataAwsIamPolicyDocument(
        resource,
        `autoscaling_assume`,
        {
          statement: [
            {
              effect: 'Allow',
              actions: ['sts:AssumeRole'],
              principals: [
                {
                  identifiers: ['ecs.application-autoscaling.amazonaws.com'],
                  type: 'Service',
                },
              ],
            },
          ],
        }
      ).json,
    });
  }

  /**
   * Creates an IAM Role Policy
   *
   * @param resource
   * @param config
   * @param iamRole
   */
  static generateIamRolePolicy(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    iamRole: iam.IamRole
  ): void {
    new IamRolePolicy(resource, `autoscaling_role_policy`, {
      name: `${config.prefix}-AutoScalingPolicy`,
      role: iamRole.id,
      policy: new DataAwsIamPolicyDocument(resource, `role_policy`, {
        statement: [
          {
            effect: 'Allow',
            actions: [
              'cloudwatch:PutMetricAlarm',
              'cloudwatch:DescribeAlarms',
              'cloudwatch:DeleteAlarms',
            ],
            resources: [
              `arn:aws:cloudwatch:*:*:alarm:/${config.ecsServiceName}*`,
            ],
          },
          {
            effect: 'Allow',
            actions: ['ecs:DescribeServices', 'ecs:UpdateService'],
            resources: [`arn:aws:ecs:*:*:service/${config.ecsServiceName}*`],
          },
        ],
      }).json,
    });
  }

  /**
   * Creates an Auto Scaling Target
   * @param resource
   * @param config
   * @param iamRole
   * @returns AppautoscalingTarget
   */
  static generateAutoScalingTarget(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    iamRole: iam.IamRole
  ): appautoscaling.AppautoscalingTarget {
    return new AppautoscalingTarget(resource, `autoscaling_target`, {
      maxCapacity: config.targetMaxCapacity,
      minCapacity: config.targetMinCapacity,
      resourceId: `service/${config.ecsClusterName}/${config.ecsServiceName}`,
      roleArn: iamRole.arn,
      scalableDimension: 'ecs:service:DesiredCount',
      serviceNamespace: 'ecs',
    });
  }

  /**
   * Creates an Autoscaling Policy
   * @param resource
   * @param config
   * @param target
   * @param type
   * @returns AppautoscalingPolicy
   */
  static generateAutoSclaingPolicy(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    target: appautoscaling.AppautoscalingTarget,
    type: 'In' | 'Out'
  ): appautoscaling.AppautoscalingPolicy {
    let stepAdjustment;

    if (type === 'In') {
      stepAdjustment = [
        {
          metricIntervalUpperBound: '0',
          scalingAdjustment: config.stepScaleInAdjustment,
        },
      ];
    } else {
      stepAdjustment = [
        {
          metricIntervalLowerBound: '0',
          scalingAdjustment: config.stepScaleOutAdjustment,
        },
      ];
    }

    return new AppautoscalingPolicy(
      resource,
      `scale_${type.toLowerCase()}_policy`,
      {
        name: `${config.prefix}-Scale${type}Policy`,
        policyType: 'StepScaling',
        resourceId: target.resourceId,
        scalableDimension: target.scalableDimension,
        serviceNamespace: target.serviceNamespace,

        stepScalingPolicyConfiguration: {
          adjustmentType: `ChangeInCapacity`,
          cooldown: 60,
          metricAggregationType: 'Average',
          stepAdjustment,
        },
        dependsOn: [target],
      }
    );
  }

  /**
   * Creates a Cloudwatch Metric Alarm
   * @param resource
   * @param config
   * @param id
   * @param name
   * @param desc
   * @param operator
   * @param threshold
   * @param arn
   */
  static generateCloudwatchMetricAlarm(
    resource: Resource,
    config: ApplicationAutoscalingProps,
    id: string,
    name: string,
    desc: string,
    operator: string,
    threshold: number,
    arn: string
  ): void {
    new CloudwatchMetricAlarm(resource, id, {
      alarmName: name,
      alarmDescription: desc,
      comparisonOperator: operator,
      evaluationPeriods: 2,
      threshold,
      statistic: 'Average',
      period: 60,
      namespace: 'AWS/ECS',
      metricName: 'CPUUtilization',
      treatMissingData: 'notBreaching',
      dimensions: {
        ClusterName: config.ecsClusterName,
        ServiceName: config.ecsServiceName,
      },
      alarmActions: [arn],
      tags: config.tags,
    });
  }
}
