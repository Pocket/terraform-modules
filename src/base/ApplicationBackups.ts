import { Construct } from 'constructs';
import { TerraformMetaArguments } from 'cdktf';
import {
  BackupPlanRule,
  BackupPlan,
} from '@cdktf/provider-aws/lib/backup-plan';
import {
  BackupSelection,
  BackupSelectionSelectionTag,
} from '@cdktf/provider-aws/lib/backup-selection';
import { BackupVault } from '@cdktf/provider-aws/lib/backup-vault';
import { BackupVaultPolicy } from '@cdktf/provider-aws/lib/backup-vault-policy';

export interface ApplicationBackupProps extends TerraformMetaArguments {
  name: string;
  kmsKeyArn: string;
  prefix: string;
  accountId: string;
  vaultPolicy: string;
  backupPlans: {
    name: string;
    resources: string[];
    rules: Omit<BackupPlanRule, 'targetVaultName'>[];
    selectionTag: BackupSelectionSelectionTag[];
  }[];
  tags?: { [key: string]: string };
}

export class ApplicationBackup extends Construct {
  public backupPlan: BackupPlan;
  public backupSelection: BackupSelection;
  public backupPlanRule: BackupPlanRule;
  private static vault: BackupVault;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationBackupProps,
  ) {
    super(scope, name);

    const vault = new BackupVault(this, 'backup-vault', {
      name: `${config.prefix}-${config.name}`,
      kmsKeyArn: config.kmsKeyArn,
      tags: config.tags,
      provider: config.provider,
    });

    new BackupVaultPolicy(this, 'backup-vault-policy', {
      backupVaultName: vault.name,
      policy: config.vaultPolicy,
      provider: config.provider,
    });

    config.backupPlans.forEach((plan) => {
      const backupPlan = new BackupPlan(this, 'backup-plan', {
        name: plan.name,
        rule: plan.rules.map((rule) => ({
          ...rule,
          targetVaultName: vault.name,
        })),
        tags: config.tags,
        provider: config.provider,
      });

      new BackupSelection(this, 'backup-selection', {
        name: `${config.prefix}-Backup-Selection`,
        planId: backupPlan.id,
        iamRoleArn: `arn:aws:iam::${config.accountId}:role/service-role/AWSBackupDefaultServiceRole`,
        resources: plan.resources,
        selectionTag: plan.selectionTag,
        provider: config.provider,
      });
    });
  }
}
