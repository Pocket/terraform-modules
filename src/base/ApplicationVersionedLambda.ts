import { Fn, Resource } from 'cdktf';
import { Construct } from 'constructs';
import { cloudwatch, iam, lambdafunction, s3 } from '@cdktf/provider-aws';
import {
  DataArchiveFile,
  DataArchiveFileSource,
} from '@cdktf/provider-archive';

export enum LAMBDA_RUNTIMES {
  PYTHON38 = 'python3.8',
  NODEJS12 = 'nodejs12.x',
  NODEJS14 = 'nodejs14.x',
}

export interface ApplicationVersionedLambdaProps {
  name: string;
  description?: string;
  runtime: LAMBDA_RUNTIMES;
  handler: string;
  timeout?: number;
  environment?: { [key: string]: string };
  vpcConfig?: lambdafunction.LambdaFunctionVpcConfig;
  executionPolicyStatements?: iam.DataAwsIamPolicyDocumentStatement[];
  tags?: { [key: string]: string };
  logRetention?: number;
  s3Bucket: string;
  usesCodeDeploy?: boolean;
}

const DEFAULT_TIMEOUT = 5;
const DEFAULT_RETENTION = 14;

export class ApplicationVersionedLambda extends Resource {
  public readonly versionedLambda: lambdafunction.LambdaAlias;
  public lambdaExecutionRole: iam.IamRole;

  constructor(
    scope: Construct,
    name: string,
    private config: ApplicationVersionedLambdaProps
  ) {
    super(scope, name);

    this.createCodeBucket();
    this.versionedLambda = this.createLambdaFunction();
  }

  private createLambdaFunction() {
    this.lambdaExecutionRole = new iam.IamRole(this, 'execution-role', {
      name: `${this.config.name}-ExecutionRole`,
      assumeRolePolicy: this.getLambdaAssumePolicyDocument(),
    });

    const executionPolicy = new iam.IamPolicy(this, 'execution-policy', {
      name: `${this.config.name}-ExecutionRolePolicy`,
      policy: this.getLambdaExecutionPolicyDocument(),
    });

    new iam.IamRolePolicyAttachment(this, 'execution-role-policy-attachment', {
      role: this.lambdaExecutionRole.name,
      policyArn: executionPolicy.arn,
      dependsOn: [this.lambdaExecutionRole, executionPolicy],
    });

    const defaultLambda = this.getDefaultLambda();
    const lambdaConfig: lambdafunction.LambdaFunctionConfig = {
      functionName: `${this.config.name}-Function`,
      filename: defaultLambda.outputPath,
      handler: this.config.handler,
      runtime: this.config.runtime,
      timeout: this.config.timeout ?? DEFAULT_TIMEOUT,
      sourceCodeHash: defaultLambda.outputBase64Sha256,
      role: this.lambdaExecutionRole.arn,
      vpcConfig: this.config.vpcConfig,
      publish: true,
      lifecycle: {
        ignoreChanges: [
          'filename',
          'source_code_hash',
          this.shouldIgnorePublish() ? 'publish' : '',
        ].filter((v: string) => v),
      },
      tags: this.config.tags,
      environment: this.config.environment
        ? { variables: this.config.environment }
        : undefined,
    };

    const lambda = new lambdafunction.LambdaFunction(
      this,
      'lambda',
      lambdaConfig
    );

    new cloudwatch.CloudwatchLogGroup(this, 'log-group', {
      name: `/aws/lambda/${lambda.functionName}`,
      retentionInDays: this.config.logRetention ?? DEFAULT_RETENTION,
      dependsOn: [lambda],
    });

    return new lambdafunction.LambdaAlias(this, 'alias', {
      functionName: lambda.functionName,
      functionVersion: Fn.element(Fn.split(':', lambda.qualifiedArn), 7),
      name: 'DEPLOYED',
      lifecycle: {
        ignoreChanges: ['function_version'],
      },
      dependsOn: [lambda],
    });
  }

  private shouldIgnorePublish() {
    if (this.config.usesCodeDeploy !== undefined) {
      return this.config.usesCodeDeploy;
    }

    return false;
  }

  private getLambdaAssumePolicyDocument() {
    return new iam.DataAwsIamPolicyDocument(this, 'assume-policy-document', {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: ['sts:AssumeRole'],
          principals: [
            {
              identifiers: ['lambda.amazonaws.com', 'edgelambda.amazonaws.com'],
              type: 'Service',
            },
          ],
        },
      ],
    }).json;
  }

  private getLambdaExecutionPolicyDocument() {
    const document = {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents',
            'logs:DescribeLogStreams',
          ],
          resources: ['arn:aws:logs:*:*:*'],
        },
        ...(this.config.executionPolicyStatements ?? []),
      ],
    };

    if (this.config.vpcConfig) {
      document.statement.push({
        effect: 'Allow',
        actions: [
          'ec2:DescribeNetworkInterfaces',
          'ec2:CreateNetworkInterface',
          'ec2:DeleteNetworkInterface',
          'ec2:DescribeInstances',
          'ec2:AttachNetworkInterface',
        ],
        resources: ['*'],
      });
    }

    return new iam.DataAwsIamPolicyDocument(
      this,
      'execution-policy-document',
      document
    ).json;
  }

  private getDefaultLambda() {
    const source = this.getDefaultLambdaSource();
    return new DataArchiveFile(this, 'lambda-default-file', {
      type: 'zip',
      source: [source],
      outputPath: `${source.filename}.zip`,
    });
  }

  private getDefaultLambdaSource(): DataArchiveFileSource {
    const runtime = this.config.runtime.match(/[a-z]*/)[0];
    const handler = this.config.handler.split('.');
    const functionName = handler.pop();
    const functionFilename = handler.join('.');

    let content = `exports.${functionName} = (event, context) => { console.log(event) }`;
    let filename = `${functionFilename}.js`;

    if (runtime === 'python') {
      content = `import json\ndef ${functionName}(event, context):\n\t print(event)\n\t return {'statusCode': 200, 'headers': {'dance': 'party'}, 'body': json.dumps({'electric': 'boogaloo'}), 'isBase64Encoded': False}`;
      filename = `${functionFilename}.py`;
    }

    return {
      content,
      filename,
    };
  }

  private createCodeBucket() {
    const codeBucket = new s3.S3Bucket(this, 'code-bucket', {
      bucket: this.config.s3Bucket,
      acl: 'private',
      tags: this.config.tags,
      forceDestroy: true,
    });

    new s3.S3BucketPublicAccessBlock(this, `code-bucket-public-access-block`, {
      bucket: codeBucket.id,
      blockPublicAcls: true,
      blockPublicPolicy: true,
    });
  }
}
