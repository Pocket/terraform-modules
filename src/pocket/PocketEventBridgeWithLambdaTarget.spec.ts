import { Testing } from 'cdktf';
import {
  PocketEventBridgeWithLambdaTarget,
  PocketEventBridgeWithLambdaTargetProps,
} from './PocketEventBridgeWithLambdaTarget';
import { LAMBDA_RUNTIMES } from '../base/ApplicationVersionedLambda';

const config: PocketEventBridgeWithLambdaTargetProps = {
  name: 'test-event-bridge-lambda',
  eventRule: {
    eventPattern: {
      source: ['aws.states'],
      'detail-type': ['Step Functions Execution Status Change'],
    },
  },
  lambda: {
    runtime: LAMBDA_RUNTIMES.PYTHON38,
    handler: 'index.handler',
  },
};

test('renders an event bridge and lambda target with rule description', () => {
  const synthed = Testing.synthScope((stack) => {
    new PocketEventBridgeWithLambdaTarget(stack, 'test-event-bridge-lambda', {
      ...config,
      eventRule: { ...config.eventRule, description: 'Test description' },
    });
  });
  expect(synthed).toMatchSnapshot();
});

test('renders an event bridge and lambda target with event bus name', () => {
  const synthed = Testing.synthScope((stack) => {
    new PocketEventBridgeWithLambdaTarget(stack, 'test-event-bridge-lambda', {
      ...config,
      eventRule: { ...config.eventRule, eventBusName: 'test-bus' },
    });
  });
  expect(synthed).toMatchSnapshot();
});
