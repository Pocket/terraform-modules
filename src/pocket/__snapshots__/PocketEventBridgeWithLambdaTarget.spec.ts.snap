// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an event bridge and lambda target with event bus name 1`] = `
"{
  \\"data\\": {
    \\"archive_file\\": {
      \\"test-event-bridge-lambda_lambda-default-file_57164BB7\\": {
        \\"output_path\\": \\"index.py.zip\\",
        \\"source\\": [
          {
            \\"content\\": \\"handler(event, context):\\\\n\\\\t print(event)\\",
            \\"filename\\": \\"index.py\\"
          }
        ],
        \\"type\\": \\"zip\\"
      }
    },
    \\"aws_iam_policy_document\\": {
      \\"test-event-bridge-lambda_assume-policy-document_59E97C0B\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"lambda.amazonaws.com\\",
                  \\"edgelambda.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      },
      \\"test-event-bridge-lambda_execution-policy-document_0A3C0E28\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"logs:CreateLogGroup\\",
              \\"logs:CreateLogStream\\",
              \\"logs:PutLogEvents\\",
              \\"logs:DescribeLogStreams\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:logs:*:*:*\\"
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      }
    }
  },
  \\"resource\\": {
    \\"aws_cloudwatch_event_rule\\": {
      \\"test-event-bridge-lambda_event-bridge-rule_529FF7C2\\": {
        \\"event_bus_name\\": \\"test-bus\\",
        \\"event_pattern\\": \\"{\\\\\\"source\\\\\\":[\\\\\\"aws.states\\\\\\"],\\\\\\"detail-type\\\\\\":[\\\\\\"Step Functions Execution Status Change\\\\\\"]}\\",
        \\"name\\": \\"test-event-bridge-lambda-Rule\\"
      }
    },
    \\"aws_cloudwatch_event_target\\": {
      \\"test-event-bridge-lambda_event-bridge-rule_event-bridge-target_84D24AD3\\": {
        \\"arn\\": \\"\${aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC.arn}\\",
        \\"depends_on\\": [
          \\"aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC\\",
          \\"aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2\\"
        ],
        \\"rule\\": \\"\${aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2.name}\\",
        \\"target_id\\": \\"lambda\\"
      }
    },
    \\"aws_cloudwatch_log_group\\": {
      \\"test-event-bridge-lambda_log-group_CF4CF8D4\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-event-bridge-lambda_65C2E1CE\\"
        ],
        \\"name\\": \\"/aws/lambda/\${aws_lambda_function.test-event-bridge-lambda_65C2E1CE.function_name}\\",
        \\"retention_in_days\\": 14
      }
    },
    \\"aws_iam_policy\\": {
      \\"test-event-bridge-lambda_execution-policy_F6422B9B\\": {
        \\"name\\": \\"test-event-bridge-lambda-ExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test-event-bridge-lambda_execution-policy-document_0A3C0E28.json}\\"
      }
    },
    \\"aws_iam_role\\": {
      \\"test-event-bridge-lambda_execution-role_7936A26A\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test-event-bridge-lambda_assume-policy-document_59E97C0B.json}\\",
        \\"name\\": \\"test-event-bridge-lambda-ExecutionRole\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test-event-bridge-lambda_execution-role-policy-attachment_2614BB3C\\": {
        \\"depends_on\\": [
          \\"aws_iam_role.test-event-bridge-lambda_execution-role_7936A26A\\",
          \\"aws_iam_policy.test-event-bridge-lambda_execution-policy_F6422B9B\\"
        ],
        \\"policy_arn\\": \\"\${aws_iam_policy.test-event-bridge-lambda_execution-policy_F6422B9B.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test-event-bridge-lambda_execution-role_7936A26A.name}\\"
      }
    },
    \\"aws_lambda_alias\\": {
      \\"test-event-bridge-lambda_alias_2EC274FC\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-event-bridge-lambda_65C2E1CE\\"
        ],
        \\"function_name\\": \\"\${aws_lambda_function.test-event-bridge-lambda_65C2E1CE.function_name}\\",
        \\"function_version\\": \\"\${element(split(\\\\\\":\\\\\\", \\\\\\"aws_lambda_function.test-event-bridge-lambda_65C2E1CE\\\\\\"), 7)}\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"function_version\\"
          ]
        },
        \\"name\\": \\"DEPLOYED\\"
      }
    },
    \\"aws_lambda_function\\": {
      \\"test-event-bridge-lambda_65C2E1CE\\": {
        \\"filename\\": \\"\${data.archive_file.test-event-bridge-lambda_lambda-default-file_57164BB7.output_path}\\",
        \\"function_name\\": \\"test-event-bridge-lambda-Function\\",
        \\"handler\\": \\"index.handler\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"filename\\",
            \\"source_code_hash\\"
          ]
        },
        \\"publish\\": true,
        \\"role\\": \\"\${aws_iam_role.test-event-bridge-lambda_execution-role_7936A26A.arn}\\",
        \\"runtime\\": \\"python3.8\\",
        \\"source_code_hash\\": \\"\${data.archive_file.test-event-bridge-lambda_lambda-default-file_57164BB7.output_base64sha256}\\",
        \\"timeout\\": 5
      }
    },
    \\"aws_lambda_permission\\": {
      \\"test-event-bridge-lambda_lambda-permission_D900B300\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"depends_on\\": [
          \\"aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC\\",
          \\"aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2\\"
        ],
        \\"function_name\\": \\"\${aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC.function_name}\\",
        \\"principal\\": \\"events.amazonaws.com\\",
        \\"qualifier\\": \\"\${aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC.name}\\",
        \\"source_arn\\": \\"\${aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2.arn}\\"
      }
    },
    \\"aws_s3_bucket\\": {
      \\"test-event-bridge-lambda_code-bucket_3703E73B\\": {
        \\"acl\\": \\"private\\",
        \\"bucket\\": \\"pocket-test-event-bridge-lambda\\",
        \\"force_destroy\\": true
      }
    },
    \\"aws_s3_bucket_public_access_block\\": {
      \\"test-event-bridge-lambda_code-bucket-public-access-block_A1A4EF9D\\": {
        \\"block_public_acls\\": true,
        \\"block_public_policy\\": true,
        \\"bucket\\": \\"\${aws_s3_bucket.test-event-bridge-lambda_code-bucket_3703E73B.id}\\"
      }
    }
  }
}"
`;

exports[`renders an event bridge and lambda target with rule description 1`] = `
"{
  \\"data\\": {
    \\"archive_file\\": {
      \\"test-event-bridge-lambda_lambda-default-file_57164BB7\\": {
        \\"output_path\\": \\"index.py.zip\\",
        \\"source\\": [
          {
            \\"content\\": \\"handler(event, context):\\\\n\\\\t print(event)\\",
            \\"filename\\": \\"index.py\\"
          }
        ],
        \\"type\\": \\"zip\\"
      }
    },
    \\"aws_iam_policy_document\\": {
      \\"test-event-bridge-lambda_assume-policy-document_59E97C0B\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"lambda.amazonaws.com\\",
                  \\"edgelambda.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      },
      \\"test-event-bridge-lambda_execution-policy-document_0A3C0E28\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"logs:CreateLogGroup\\",
              \\"logs:CreateLogStream\\",
              \\"logs:PutLogEvents\\",
              \\"logs:DescribeLogStreams\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:logs:*:*:*\\"
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      }
    }
  },
  \\"resource\\": {
    \\"aws_cloudwatch_event_rule\\": {
      \\"test-event-bridge-lambda_event-bridge-rule_529FF7C2\\": {
        \\"description\\": \\"Test description\\",
        \\"event_bus_name\\": \\"default\\",
        \\"event_pattern\\": \\"{\\\\\\"source\\\\\\":[\\\\\\"aws.states\\\\\\"],\\\\\\"detail-type\\\\\\":[\\\\\\"Step Functions Execution Status Change\\\\\\"]}\\",
        \\"name\\": \\"test-event-bridge-lambda-Rule\\"
      }
    },
    \\"aws_cloudwatch_event_target\\": {
      \\"test-event-bridge-lambda_event-bridge-rule_event-bridge-target_84D24AD3\\": {
        \\"arn\\": \\"\${aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC.arn}\\",
        \\"depends_on\\": [
          \\"aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC\\",
          \\"aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2\\"
        ],
        \\"rule\\": \\"\${aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2.name}\\",
        \\"target_id\\": \\"lambda\\"
      }
    },
    \\"aws_cloudwatch_log_group\\": {
      \\"test-event-bridge-lambda_log-group_CF4CF8D4\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-event-bridge-lambda_65C2E1CE\\"
        ],
        \\"name\\": \\"/aws/lambda/\${aws_lambda_function.test-event-bridge-lambda_65C2E1CE.function_name}\\",
        \\"retention_in_days\\": 14
      }
    },
    \\"aws_iam_policy\\": {
      \\"test-event-bridge-lambda_execution-policy_F6422B9B\\": {
        \\"name\\": \\"test-event-bridge-lambda-ExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test-event-bridge-lambda_execution-policy-document_0A3C0E28.json}\\"
      }
    },
    \\"aws_iam_role\\": {
      \\"test-event-bridge-lambda_execution-role_7936A26A\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test-event-bridge-lambda_assume-policy-document_59E97C0B.json}\\",
        \\"name\\": \\"test-event-bridge-lambda-ExecutionRole\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test-event-bridge-lambda_execution-role-policy-attachment_2614BB3C\\": {
        \\"depends_on\\": [
          \\"aws_iam_role.test-event-bridge-lambda_execution-role_7936A26A\\",
          \\"aws_iam_policy.test-event-bridge-lambda_execution-policy_F6422B9B\\"
        ],
        \\"policy_arn\\": \\"\${aws_iam_policy.test-event-bridge-lambda_execution-policy_F6422B9B.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test-event-bridge-lambda_execution-role_7936A26A.name}\\"
      }
    },
    \\"aws_lambda_alias\\": {
      \\"test-event-bridge-lambda_alias_2EC274FC\\": {
        \\"depends_on\\": [
          \\"aws_lambda_function.test-event-bridge-lambda_65C2E1CE\\"
        ],
        \\"function_name\\": \\"\${aws_lambda_function.test-event-bridge-lambda_65C2E1CE.function_name}\\",
        \\"function_version\\": \\"\${element(split(\\\\\\":\\\\\\", \\\\\\"aws_lambda_function.test-event-bridge-lambda_65C2E1CE\\\\\\"), 7)}\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"function_version\\"
          ]
        },
        \\"name\\": \\"DEPLOYED\\"
      }
    },
    \\"aws_lambda_function\\": {
      \\"test-event-bridge-lambda_65C2E1CE\\": {
        \\"filename\\": \\"\${data.archive_file.test-event-bridge-lambda_lambda-default-file_57164BB7.output_path}\\",
        \\"function_name\\": \\"test-event-bridge-lambda-Function\\",
        \\"handler\\": \\"index.handler\\",
        \\"lifecycle\\": {
          \\"ignore_changes\\": [
            \\"filename\\",
            \\"source_code_hash\\"
          ]
        },
        \\"publish\\": true,
        \\"role\\": \\"\${aws_iam_role.test-event-bridge-lambda_execution-role_7936A26A.arn}\\",
        \\"runtime\\": \\"python3.8\\",
        \\"source_code_hash\\": \\"\${data.archive_file.test-event-bridge-lambda_lambda-default-file_57164BB7.output_base64sha256}\\",
        \\"timeout\\": 5
      }
    },
    \\"aws_lambda_permission\\": {
      \\"test-event-bridge-lambda_lambda-permission_D900B300\\": {
        \\"action\\": \\"lambda:InvokeFunction\\",
        \\"depends_on\\": [
          \\"aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC\\",
          \\"aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2\\"
        ],
        \\"function_name\\": \\"\${aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC.function_name}\\",
        \\"principal\\": \\"events.amazonaws.com\\",
        \\"qualifier\\": \\"\${aws_lambda_alias.test-event-bridge-lambda_alias_2EC274FC.name}\\",
        \\"source_arn\\": \\"\${aws_cloudwatch_event_rule.test-event-bridge-lambda_event-bridge-rule_529FF7C2.arn}\\"
      }
    },
    \\"aws_s3_bucket\\": {
      \\"test-event-bridge-lambda_code-bucket_3703E73B\\": {
        \\"acl\\": \\"private\\",
        \\"bucket\\": \\"pocket-test-event-bridge-lambda\\",
        \\"force_destroy\\": true
      }
    },
    \\"aws_s3_bucket_public_access_block\\": {
      \\"test-event-bridge-lambda_code-bucket-public-access-block_A1A4EF9D\\": {
        \\"block_public_acls\\": true,
        \\"block_public_policy\\": true,
        \\"bucket\\": \\"\${aws_s3_bucket.test-event-bridge-lambda_code-bucket_3703E73B.id}\\"
      }
    }
  }
}"
`;
